File: Java/Lesson075_MergeKSortedLists.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 075 -- Merge K Sorted Lists
// Category   : Linked List
// Difficulty : Hard
// Study Plan : Day 38
// =============================================================
//
// QUESTION:
//   Merge k sorted linked lists into one sorted list.
//
//   Example: [[1,4,5],[1,3,4],[2,6]] -> 1->1->2->3->4->4->5->6
// =============================================================

import java.util.*;
public class Lesson075_MergeKSortedLists {
    static class ListNode { int val; ListNode next; ListNode(int v){val=v;} }
    public ListNode mergeKLists(ListNode[] lists){
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a,b)->a.val-b.val);
        for (ListNode n: lists) if (n!=null) pq.add(n);
        ListNode d = new ListNode(0), cur = d;
        while (!pq.isEmpty()){
            ListNode n = pq.poll(); cur.next = n; cur = n;
            if (n.next != null) pq.add(n.next);
        }
        return d.next;
    }
    static ListNode fromArr(int[] a){ ListNode d=new ListNode(0), c=d; for (int x: a){ c.next=new ListNode(x); c=c.next; } return d.next; }
    static List<Integer> toArr(ListNode n){ List<Integer> r=new ArrayList<>(); while (n!=null){ r.add(n.val); n=n.next; } return r; }
    public static void main(String[] a){
        ListNode[] ls = { fromArr(new int[]{1,4,5}), fromArr(new int[]{1,3,4}), fromArr(new int[]{2,6}) };
        System.out.println(toArr(new Lesson075_MergeKSortedLists().mergeKLists(ls)));
    }
}

File: CPP/Lesson_075_merge_k_sorted_lists.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 075 -- Merge K Sorted Lists
// Category   : Linked List
// Difficulty : Hard
// Study Plan : Day 38
// =============================================================
//
// QUESTION:
//   Merge k sorted linked lists into one sorted list.
//
//   Example: [[1,4,5],[1,3,4],[2,6]] -> 1->1->2->3->4->4->5->6
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
struct ListNode { int val; ListNode* next; ListNode(int v):val(v),next(nullptr){} };
struct Cmp { bool operator()(ListNode* a, ListNode* b) const { return a->val > b->val; } };
class Solution { public:
    ListNode* mergeKLists(vector<ListNode*>& lists){
        priority_queue<ListNode*, vector<ListNode*>, Cmp> pq;
        for (auto n: lists) if (n) pq.push(n);
        ListNode d(0), *cur = &d;
        while (!pq.empty()){
            auto n = pq.top(); pq.pop(); cur->next = n; cur = n;
            if (n->next) pq.push(n->next);
        }
        return d.next;
    }
};
ListNode* fromArr(vector<int> a){ ListNode d(0), *c=&d; for (int x: a){ c->next=new ListNode(x); c=c->next; } return d.next; }
int main(){ vector<ListNode*> ls = { fromArr({1,4,5}), fromArr({1,3,4}), fromArr({2,6}) };
    auto r = Solution().mergeKLists(ls); while (r){ cout<<r->val<<" "; r=r->next; } cout<<endl; }
